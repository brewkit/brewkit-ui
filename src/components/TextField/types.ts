import * as React from 'react';
import { Theme } from '@material-ui/core/styles/createMuiTheme';
import { TextFieldProps } from '@material-ui/core/TextField';


/**
 * Declare any theme variables we want to make available.
 */
declare module '@material-ui/core/styles/createMuiTheme' {
    interface Theme {
        BkTextField: {
            searchIcon: React.CSSProperties,
            visibilityIcon: React.CSSProperties,
            menu: React.CSSProperties,
            menuItem: React.CSSProperties,
            activeMenuItem: React.CSSProperties,
        },
    }
    // allow configuration using `createMuiTheme`
    interface ThemeOptions {
        BkTextField?: {
            searchIcon?: React.CSSProperties,
            visibilityIcon?: React.CSSProperties,
            menu?: React.CSSProperties,
            menuItem?: React.CSSProperties,
            activeMenuItem?: React.CSSProperties,
        },
    }
}


export type Props = TextFieldProps & {

    options?: {
        label: React.ReactNode,
        value: any,
    }[],

    /**
     * @ignore
     */
    classes?: Record<string, any>,

};
