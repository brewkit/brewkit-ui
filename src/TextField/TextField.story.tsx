import * as React from 'react';
import MenuItem from '@material-ui/core/MenuItem';
import { TextField } from './TextField';

export default {
    component: TextField,
    parameters: {
        // componentSubtitle: <Flag color="success">Stable</Flag>,
    },
    title: 'Inputs/TextField',
};

const defaultArgs = {
    disabled: false,
    error: false,
    fullWidth: false,
    helperText: 'Some helper text',
    label: 'Form Label',
    multiline: false,
    placeholder: 'Some placeholder text',
};

export const Sandbox = (args: any): React.ReactElement => (
    <TextField {...args} />
);
Sandbox.args = {
    ...defaultArgs,
};

export const Types = (): React.ReactElement => (
    <div style={{ display: 'grid', gridGap: '1rem' }}>
        <TextField label="Text" type="text" />
        <TextField label="Password" type="password" />
        <TextField label="Search" type="search" />
        <TextField label="Number" type="number" />
        <TextField
            defaultValue="Default Value"
            label="Multiline"
            multiline
            rows={4}
        />
        <TextField
            helperText="Please select your number."
            id="outlined-select-currency"
            label="Select"
            select
            variant="outlined">
            <MenuItem value="One">One</MenuItem>
            <MenuItem value="Two">Two</MenuItem>
            <MenuItem value="Three">Three</MenuItem>
        </TextField>
    </div>
);

export const Select = (): React.ReactElement => {
    const options = [
        { label: '', value: '', hidden: true },
        { label: 'one', value: 1 },
        { label: 'two', value: 2 },
        { label: 'three', value: 3 },
        { label: 'four', value: 4 },
    ];

    return (
        <TextField
            defaultValue=""
            helperText="Please select your number."
            label="Select"
            options={options}
            select
            variant="outlined"
        />
    );
};
