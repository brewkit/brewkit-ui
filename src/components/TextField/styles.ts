import { Theme } from '@material-ui/core/styles/createMuiTheme';
import { merge as _merge } from 'lodash';


const styles = (theme: Theme): any => ({


    searchIcon: _merge({
        display: 'flex',
        marginRight: '0.5rem',
        color: theme?.palette?.primary?.main,
    }, theme?.BkTextField?.searchIcon),


    visibilityIcon: _merge({
        display: 'flex',
        marginLeft: '0.5rem',
        color: theme?.palette?.primary?.main,
        cursor: 'pointer',
    }, theme?.BkTextField?.visibilityIcon),


    menu: _merge({}, theme?.BkTextField?.menu),

    menuItem: _merge({}, theme?.BkTextField?.menuItem),

    activeMenuItem: _merge({}, theme?.BkTextField?.activeMenuItem),


});


export default styles;
