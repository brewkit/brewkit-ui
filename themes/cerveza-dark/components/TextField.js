import { color } from '../../cerveza/variables';
var OutlinedInput = {
    overrides: {
        MuiFormHelperText: {
            root: {
                color: color.gray.light
            }
        }
    }
};
export default OutlinedInput;
