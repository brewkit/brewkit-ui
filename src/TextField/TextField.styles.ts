import { Theme } from '@material-ui/core/styles/createTheme';
import { createStyles, makeStyles } from '@material-ui/core/styles';
import merge from 'lodash.merge';

declare module '@material-ui/core/styles/createTheme' {
    interface Theme {
        BkTextField: {
            searchIcon: React.CSSProperties;
            visibilityIcon: React.CSSProperties;
            menu: React.CSSProperties;
            menuItem: React.CSSProperties;
            activeMenuItem: React.CSSProperties;
        };
    }

    interface ThemeOptions {
        BkTextField?: {
            searchIcon?: React.CSSProperties;
            visibilityIcon?: React.CSSProperties;
            menu?: React.CSSProperties;
            menuItem?: React.CSSProperties;
            activeMenuItem?: React.CSSProperties;
        };
    }
}

export const useMuiStyles = makeStyles(() =>
    createStyles({
        root: {
            '& label': {
                transform: 'translate(18px, 18px) scale(1)',
            },
        },
    }),
);

export default makeStyles(
    (theme) =>
        createStyles({
            searchIcon: merge(
                {
                    display: 'flex',
                    marginRight: '0.5rem',
                    color: theme?.palette?.primary?.main,
                },
                theme?.BkTextField?.searchIcon,
            ),

            visibilityIcon: merge(
                {
                    display: 'flex',
                    marginLeft: '0.5rem',
                    color: theme?.palette?.primary?.main,
                    cursor: 'pointer',
                },
                theme?.BkTextField?.visibilityIcon,
            ),

            menu: merge({}, theme?.BkTextField?.menu),

            menuItem: merge({}, theme?.BkTextField?.menuItem),

            activeMenuItem: merge({}, theme?.BkTextField?.activeMenuItem),
        }),
    { name: 'TextField' },
);
