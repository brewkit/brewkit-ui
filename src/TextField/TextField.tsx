import * as React from 'react';
import MUITextField, {
    TextFieldClassKey,
    TextFieldProps as MUITextFieldProps,
} from '@material-ui/core/TextField';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import { useTheme } from '@material-ui/core/styles';
import MenuItem from '@material-ui/core/MenuItem';
import Visibility from '@material-ui/icons/VisibilityOutlined';
import VisibilityOff from '@material-ui/icons/VisibilityOffOutlined';
import Search from '@material-ui/icons/SearchOutlined';
import { motion, AnimatePresence } from 'framer-motion';
import { merge } from 'lodash';
import { ClassNameMap } from '@material-ui/styles';
import { SelectProps as MuiSelectProps } from '@material-ui/core';
import clsx from 'clsx';
import withoutKeys from '../utils/withoutKeys';
import useStyles from './TextField.styles';

export type BkTextFieldProps<T = string | number> = MUITextFieldProps & {
    options?: {
        label: React.ReactNode;
        value: T;
        hidden?: boolean;
    }[];

    classes?: ClassNameMap<
        | TextFieldClassKey
        | 'activeMenuItem'
        | 'menu'
        | 'input'
        | 'menuItem'
        | 'searchIcon'
        | 'visibilityIcon'
    >;

    /**
     * Props for Select component if 'select' is true
     * @default {}
     */
    SelectProps?: MuiSelectProps;
};

const TextField = React.forwardRef(
    (
        props: BkTextFieldProps,
        ref: React.Ref<HTMLInputElement>,
    ): React.ReactElement => {
        const {
            classes: userClasses = {},
            children,
            fullWidth = true,
            type = 'text',
            select,
            options,
            InputProps = {},
            SelectProps = {},
            InputLabelProps = {},
            variant = 'outlined',
            ...otherProps
        } = props;
        const isPassword = type === 'password';
        const classes = useStyles({ isPassword });
        const muiClasses = merge(
            withoutKeys(classes, [
                'activeMenuItem',
                'menu',
                'menuItem',
                'input',
                'searchIcon',
                'visibilityIcon',
            ]),
            userClasses,
        );
        const [isVisible, setVisible] = React.useState(type !== 'password');
        const { breakpoints } = useTheme();
        const isMobile = useMediaQuery(breakpoints.down('md'));

        /**
         * figure our what type to our input is based on password visibility toggle
         */
        function getType(): string {
            return !isPassword ? type : isVisible ? 'text' : 'password';
        }

        /**
         * Figure out which starting adornment to use based on input type
         */
        function getStartAdornment(): React.ReactNode {
            if (type === 'search')
                return (
                    <span className={classes?.searchIcon}>
                        <Search />
                    </span>
                );
            return null;
        }

        /**
         * configure our Framer animation
         */
        const fadeAnim = {
            initial: { scale: 0 },
            animate: { scale: 1 },
            exit: { scale: 0 },
        };

        /**
         * generate select options from options prop. Will use MenuItem if desktop and native treatments if mobile.
         */
        function getOptions(): React.ReactNode {
            const OptionComponent = isMobile ? 'option' : MenuItem;
            return options?.map((option) => {
                if (!isMobile && option.hidden) return null;
                return (
                    <OptionComponent
                        classes={
                            isMobile
                                ? undefined
                                : {
                                      root: classes?.menuItem,
                                      selected: classes?.activeMenuItem,
                                  }
                        }
                        hidden={option.hidden}
                        key={option.value}
                        value={option.value}>
                        {option.label}
                    </OptionComponent>
                );
            });
        }

        /**
         * Figure out which end adornment to use based on input type and error state
         */
        function getEndAdornment(): React.ReactNode {
            if (type === 'password')
                return (
                    <span
                        className={classes?.visibilityIcon}
                        data-testid="visibilityIcon"
                        onClick={(): void => setVisible(!isVisible)}>
                        <AnimatePresence>
                            {isVisible ? (
                                <Visibility
                                    component={motion.svg}
                                    {...fadeAnim}
                                />
                            ) : (
                                <VisibilityOff
                                    component={motion.svg}
                                    {...fadeAnim}
                                />
                            )}
                        </AnimatePresence>
                    </span>
                );
            return null;
        }

        /**
         * set props to be passed to our underlying <Input />
         */
        const andormentProps = {
            startAdornment: getStartAdornment(),
            endAdornment: getEndAdornment(),
        };

        return (
            <MUITextField
                classes={muiClasses}
                fullWidth={fullWidth}
                InputProps={{
                    ...andormentProps,
                    ...(InputProps || {}),
                    classes: {
                        ...(InputProps?.classes || {}),
                        input: clsx(classes.input, InputProps?.classes?.input),
                    },
                }}
                InputLabelProps={{
                    shrink: true,
                    classes: {
                        ...(InputLabelProps?.classes || {}),
                    },
                    ...InputLabelProps,
                }}
                SelectProps={{
                    native: Boolean(select && isMobile) || undefined,
                    MenuProps: { classes: { paper: classes?.menu } },
                    ...SelectProps,
                }}
                ref={ref}
                select={select}
                type={getType()}
                variant={variant}
                {...otherProps}>
                {options && select ? getOptions() : children}
            </MUITextField>
        );
    },
);

TextField.displayName = 'BkTextField';

export default TextField;
