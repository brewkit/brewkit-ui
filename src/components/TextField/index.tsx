import * as React from 'react';
import MuiTextField from '@material-ui/core/TextField';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import { useTheme, withStyles } from '@material-ui/core/styles';
import MenuItem from '@material-ui/core/MenuItem';
import Visibility from '@material-ui/icons/VisibilityOutlined';
import VisibilityOff from '@material-ui/icons/VisibilityOffOutlined';
import Search from '@material-ui/icons/SearchOutlined';
import { motion, AnimatePresence } from 'framer-motion';
import { Props } from './types';
import styles from './styles';


/**
 * Text fields let users enter and edit text.
 *
 * [Material-UI Docs](https://material-ui.com/components/text-fields/)
 *
 * ## Differences in Brewkit:
 *
 * - Adds a visibility toggle when `type="password"`. Styles can be overridden using `BkTextField.visibilityIcon`.
 * - Adds a search icon when `type="search"`. Styles can be overridden using `BkTextField.searchIcon`.
 * - Adds a access to override the Select Menu and MenuItems. Styles can be overridden using `BkTextField.menu`,
 * `BkTextField.menuItem` and `BkTextField.activeMenuItem`.
 * - Adds an `options` property that can be used to pass options if `type="select"`. This is the preferred way of
 * implementing selects, as it will automatically provide a native solution at mobile resolutions.
 */
export const TextField = React.forwardRef(({
    children,
    classes,
    type = 'text',
    select,
    options,
    ...otherProps
}: Props, ref: React.Ref<any>): React.ReactElement => {


    const [isVisible, setVisible] = React.useState(type !== 'password');
    const theme = useTheme();
    const isMobile = useMediaQuery(theme.breakpoints.down('md'));


    /**
     * figure our what type to our input is based on password visibility toggle
     */
    function getType(): string {
        if (type !== 'password') return type;
        return isVisible ? 'text' : 'password';
    }


    /**
     * Figure out which starting adornment to use based on input type
     */
    function getStartAdornment(): React.ReactNode {
        if (type === 'search') return (
            <span className={classes?.searchIcon}>
                <Search />
            </span>
        );
        return null;
    }


    /**
     * configure our Framer animation
     */
    const fadeAnim = {
        initial: { scale: 0 },
        animate: { scale: 1 },
        exit: { scale: 0 },
    };


    /**
     * generate select options from options prop. Will use MenuItem if desktop and native treatments if mobile.
     */
    function getOptions(): React.ReactNode {
        const OptionComponent = isMobile ? 'option' : MenuItem;
        return options?.map((option) => (
            <OptionComponent
                classes={(isMobile ? undefined : { root: classes?.menuItem, selected: classes?.activeMenuItem })}
                key={option.value}
                value={option.value}
            >
                {option.label}
            </OptionComponent>
        ));
    }


    /**
     * Figure out which end adornment to use based on input type and error state
     */
    function getEndAdornment(): React.ReactNode {
        if (type === 'password') return (
            <span className={classes?.visibilityIcon} onClick={(): void => setVisible(!isVisible)}>
                <AnimatePresence>
                    { isVisible ? (
                        <Visibility component={motion.svg} {...fadeAnim} />
                    ) : (
                        <VisibilityOff component={motion.svg} {...fadeAnim} />
                    )}
                </AnimatePresence>
            </span>
        );
        return null;
    }


    /**
     * set props to be passed to our underlying <Input />
     */
    const InputProps = {
        startAdornment: getStartAdornment(),
        endAdornment: getEndAdornment(),
    };


    return (
        <MuiTextField
            InputProps={InputProps}
            SelectProps={{
                native: Boolean(select && isMobile) || undefined,
                MenuProps: { classes: { paper: classes?.menu } },
            }}
            ref={ref}
            select={select}
            type={getType()}
            {...otherProps}
        >
            {options && select ? getOptions() : children}
        </MuiTextField>
    );


});


TextField.displayName = 'TextField';


export default withStyles(styles)(TextField);
