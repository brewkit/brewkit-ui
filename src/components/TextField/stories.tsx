import * as React from 'react';
import { text, boolean } from '@storybook/addon-knobs';
import { withStyles } from '@material-ui/core';
import MenuItem from '@material-ui/core/MenuItem';


/**
 * This is an annoying hack to get Storybook to properly parse the documentation when the underlying component
 * is wrapped in a HOC. Expect to do/see this on every component with MUI theming options.
 */
import { TextField as Component } from './index';
import styles from './styles';
const TextField = withStyles(styles)(Component)


export default {
    component: TextField,
    parameters: {
        // componentSubtitle: <Flag color="success">Stable</Flag>,
    },
    title: 'Inputs/TextField',
};


export const Sandbox = (): React.ReactElement => {

    const label = text('label', 'Form Label');
    const placeholder = text('placeholder', 'Some placeholder text');
    const helperText = text('helperText', 'Some helper text');
    const error = boolean('error', false);
    const fullWidth = boolean('fullWidth', false);
    const disabled = boolean('disabled', false);
    const multiline = boolean('multiline', false);

    return (
        <TextField
            error={error}
            label={label}
            placeholder={placeholder}
            disabled={disabled}
            fullWidth={fullWidth}
            helperText={helperText}
            multiline={multiline}
        />
    );

};


export const Types = (): React.ReactElement => {

    return (
        <div style={{ display: 'grid', gridGap: '1rem' }}>
            <TextField label="Text" type="text" />
            <TextField label="Password" type="password" />
            <TextField label="Search" type="search" />
            <TextField label="Number" type="number" />
            <TextField
                label="Multiline"
                multiline
                rows={4}
                defaultValue="Default Value"
            />
            <TextField
                id="outlined-select-currency"
                select
                label="Select"
                helperText="Please select your number."
                variant="outlined"
            >
                <MenuItem value="One">One</MenuItem>
                <MenuItem value="Two">Two</MenuItem>
                <MenuItem value="Three">Three</MenuItem>
            </TextField>
        </div>
    );

}


export const Select = (): React.ReactElement => {

    const options = [
        { label: 'one', value: 1 },
        { label: 'two', value: 2 },
        { label: 'three', value: 3 },
        { label: 'four', value: 4 },
    ]

    return (
        <TextField
            defaultValue={2}
            select
            label="Select"
            helperText="Please select your number."
            variant="outlined"
            options={options}
        />
    );

}
