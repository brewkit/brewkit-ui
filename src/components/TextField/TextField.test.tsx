// @ts-nocheck
import * as React from 'react';
import { create } from 'react-test-renderer';
import TextField from '.';

it('Render TextField component', () => {
    const comp = create(
        <TextField value="Text data" />
    ).toJSON();

    expect(comp).toMatchSnapshot();
});
