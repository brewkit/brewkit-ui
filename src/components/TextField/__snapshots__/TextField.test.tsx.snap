// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render TextField component 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiInput-input"
      disabled={false}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value="Text data"
    />
  </div>
</div>
`;
